src/server.js

```diff
@@ -1,168 +1,194 @@
+  const { format = defaultFormat } = this.request.query;
 
-    if (supportedFormats.indexOf(format.toLowerCase()) === -1) {
-        this.throw(`Format ${format} not supported.`, 400);
-    }
+  if (supportedFormats.indexOf(format.toLowerCase()) === -1) {
+    this.throw(400, `Format ${format} not supported.`);
+  }
 
-    this.type = this.state.format = format;
-    yield next;
-});
+  this.type = this.state.format = format;
 
-/**
- * Set up the size of the page to render, either the paper size for PDF or
- * clip rectangle for image formats.
- */
-app.use(function *(next) {
-    if (this.state.format === 'pdf') {
-        yield this.state.page.property('paperSize', {
-            format: 'A4',
-            orientation: 'portrait',
-            border: '1cm',
-        });
-    }
-    else {
-        yield this.state.page.property('viewportSize')
-            .then(size => ({
-                top: 0,
-                left: 0,
-                width: size.width,
-                height: size.height,
-            }))
-            .then(clipRect => this.state.page.property('clipRect', clipRect));
-    }
-
-    yield next;
+  yield next;
 });
 
 /**
  * Generate a screenshot of the loaded page.
  *
  * If successful the screenshot is sent as the response.
  */
-app.use(function *(next) {
-    const url = this.request.query.url;
-    const format = this.state.format;
-
-    logger.info(`Rendering screenshot of ${url} to ${format}`);
-
-    if (format === 'pdf') {
-        const tmpFile = tmp.fileSync({ postfix: '.pdf'});
-
-        yield this.state.page.render(tmpFile.name)
-            .then(() => {
-                this.body = fs.createReadStream(tmpFile.name);
-
-                // Delete the temp file after served to client
-                this.body.on('close', () => tmpFile.removeCallback());
-            });
-    }
-    else {
-        yield this.state.page.renderBase64(format)
-            .then((imageData) => this.body = Buffer.from(imageData, 'base64'));
-    }
+app.use(function*(next) {
+  const { url } = this.request.query;
+  const { format, page, browser } = this.state;
+  const { width, height } = page.viewport();
+  let renderError;
+
+  logger.info(`Rendering screenshot of ${url} to ${format}`);
+
+  if (format === 'pdf') {
+    yield page
+      .pdf({
+        format: 'A4',
+        margin: { top: '1cm', right: '1cm', bottom: '1cm', left: '1cm' },
+      })
+      .then(response => (this.body = response))
+      .catch(error => (renderError = error));
+  } else {
+    yield page
+      .screenshot({
+        type: format === 'jpg' ? 'jpeg' : format,
+        clip: {
+          x: 0,
+          y: 0,
+          width: width,
+          height: height,
+        },
+        omitBackground: true,
+      })
+      .then(response => (this.body = response))
+      .catch(error => (renderError = error));
+  }
+
+  if (renderError) {
+    this.throw(400, `Could not render page: ${renderError.message}`);
+  }
+
+  yield page.close();
+});
```

Comment:
After reading up on Koa middleware again recently, I think best practice is to end with `yield next` in the last middleware too. Koa injects a noop next at the end, so it won't do anything, but it means this middleware could be positioned wherever in the chain, eg. adding another middleware after this wouldn't require us to update this middleware at all.